segmentation/utils/spatial_analysis_utils.py

```diff
@@ -250,3 +250,55 @@ def calculate_enrichment_stats(close_num, close_num_rand):
     dims = ["stats", "rows", "cols"]
     stats_xr = xr.DataArray(stats_data, coords=coords, dims=dims)
     return stats_xr
+
+
+def compute_neighbor_counts(fov_data, dist_matrix, distlim, pheno_num,
+                            cell_neighbor_counts, cell_neighbor_freqs, cell_count):
+    """Calculates the number of neighbor phenotypes for each cell
+
+    Args:
+        fov_data: data for the current fov
+        dist_matrix: cells x cells matrix with the euclidian
+            distance between centers of corresponding cells
+        distlim: threshold for spatial enrichment distance proximity
+        pheno_num: number of different cell phenotypes
+        cell_neighbor_counts: matrix with phenotype counts per cell
+        cell_neighbor_freqs: matrix with phenotype frequencies of
+            counts per phenotype/total phenotypes for each cell
+        cell_count: current cell in analysis
+    Returns:
+        cell_neighbor_counts: matrix with phenotype counts per cell
+        cell_neighbor_freqs: matrix with phenotype frequencies of
+            counts per phenotype/total phenotypes for each cell
+        cell_count: current cell in analysis"""
+
+    cell_label_col = "cellLabelInImage"
+
+    for j in list(fov_data.index):
+        # get specific cell
+        cell = fov_data.loc[j, cell_label_col]
+        # get all cell neighbors within threshold distance
+        cell_dist_mat = dist_matrix[int(cell - 1), :]
+        cell_dist_mat_bin = np.zeros(cell_dist_mat.shape)
+        # Binarize the cell_dist_mat, making all those within the distlim equal to 1
+        cell_dist_mat_bin[cell_dist_mat < distlim] = 1
+
+        # get indices (labels) of close cells
+        neighbor_labels = np.argwhere(cell_dist_mat_bin > 0) + 1
+        # filter out non-cellular objects (with no corresponding label in fov data)
+        neighbor_labels_cells = np.intersect1d(neighbor_labels, fov_data[cell_label_col])
+
+        # count phenotypes in cell neighbors
+        count_vec = np.zeros((1, pheno_num))
+        # Only include the neighbor labels that are cell labels
+        neighbor_inds = np.isin(fov_data[cell_label_col], neighbor_labels_cells)
+        # Get the phenotypes of the neighbor labels by index
+        pheno_vec = fov_data.iloc[neighbor_inds, 2]
+        for k in range(1, pheno_num):
+            count_vec[0, k - 1] = sum(pheno_vec == k)
+        # add to neighborhood matrices
+        cell_neighbor_counts.loc[cell_count, 2:] = count_vec[0, :]
+        cell_neighbor_freqs.loc[cell_count, 2:] = (count_vec[0, :] / len(neighbor_labels_cells))
+        # update cell counts
+        cell_count += 1
```

Comment:
You can actually do this with no for loop.
```
# remove non-cell2cell distances
cell_dist_mat = np.take(dist_matrix, fov_data[cell_label_col] - 1, 0)
cell_dist_mat = np.take(cell_dist_mat, fov_data[cell_label_col] - 1, 1)

# binarize distance matrix
cell_dist_mat_bin = np.zeros(cell_dist_mat.shape)
cell_dist_mat_bin[cell_dist_mat < dist_lime] = 1

# get num_neighbors for freqs
num_neighbors = np.sum(cell_dist_mat_bin, axis=0)

# create the 'phenotype has cell?' matrix, excluding non cell-label rows (should match up w/ dist_matrix)
pheno_has_cell = pd.get_dummies(fov_data.iloc[fov_data[cell_label_col], 2]).to_numpy().T

# dot binarized 'is neighbor?' matrix with pheno_has_cell to get counts
counts = pheno_has_cell.dot(dist_mat).T

# compute freqs with num_neighbors (one of these or both might need to be transposed)
freqs = counts/num_neighbors

# add to neighbor counts + freqs
cell_neighbor_counts.loc[cell_count:(cell_count+cell_dist_mat.shape[0]), 2:] = counts
cell_neighbor_freqs.loc[cell_count:(cell_count+cell_dist_mat.shape[0]), 2:] freqs
```

there are probably some array-shape related errors in the above sample/snippit (particularly with transposes, etc), but this approach, in principle, should work and should significantly speed the function up.  

basically, it uses `fov_data` to construct a binarized, "does cell k, have phenotype j?" matrix, P_jk, for all labelled-cells.  just taking one cell's binarized "is cell k my neighbor?" vector, d, it's clear that the matrix (dot in np) product P_jk*d_k = c_j is the vector containing the number of neighbors of each phenotype, j.  luckily, numpy's `dot` automatically repeats this over for each column d_k of `cell_dist_mat_bin`, which yields the counts for each cell.  
The rest is just filing away the results into `cell_neighbor_counts` and doing a little extra hoop-jumping for the frequencies.
Feel free to slack me if any of that wasn't clear, or if I'm mondo wrong